components/external_storage/src/grpc_client.rs

```diff
@@ -0,0 +1,132 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crate::request::{
+    anyhow_to_io_log_error, file_name_for_write, restore_sender, write_sender, DropPath,
+};
+use crate::ExternalStorage;
+
+use anyhow::Context;
+use futures_io::AsyncRead;
+use grpcio::{self};
+use kvproto::backup as proto;
+#[cfg(feature = "prost-codec")]
+pub use kvproto::backup::storage_backend::Backend;
+#[cfg(feature = "protobuf-codec")]
+pub use kvproto::backup::StorageBackend_oneof_backend as Backend;
+use std::io::{self, ErrorKind};
+use std::sync::Arc;
+use tikv_util::time::Limiter;
+use tokio::runtime::{Builder, Runtime};
+
+struct ExternalStorageClient {
+    backend: Backend,
+    runtime: Arc<Runtime>,
+    rpc: proto::ExternalStorageClient,
+    name: &'static str,
+    url: url::Url,
+}
+
+pub fn new_client(
+    backend: Backend,
+    name: &'static str,
+    url: url::Url,
+) -> io::Result<Box<dyn ExternalStorage>> {
+    let runtime = Builder::new()
+        .basic_scheduler()
+        .thread_name("external-storage-grpc-client")
+        .core_threads(1)
+        .enable_all()
+        .build()?;
+    Ok(Box::new(ExternalStorageClient {
+        backend,
+        runtime: Arc::new(runtime),
+        rpc: new_rpc_client()?,
+        name,
+        url,
+    }))
+}
+
+fn new_rpc_client() -> io::Result<proto::ExternalStorageClient> {
+    let env = Arc::new(grpcio::EnvBuilder::new().build());
+    let grpc_socket_path = "/tmp/grpc-external-storage.sock";
```

Comment:
What about putting it into tikv's `data-dir` to prevent being deleted unexpectedly? 

Have you considered security for the approach?  Does it require mTLS?

Reply:
yes, data-dir might be a a better place. This would operate on the same node, so TLS is less of a concern. More important for security would be securing access to the socket.

The GRPC API is experimental and I would prefer to remove it in the future and make the case for the C dynamic lib instead. So I would prefer not to spend more time on it now.